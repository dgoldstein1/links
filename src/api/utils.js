import axios from "axios";
// makes request
// returns Promise{sucess, data, error}
export function makeRequest({ method, url, onErrorPrefix = "", body }) {
  return axios[`${method}`](encodeURI(url), body)
    .then(r => ({
      success: true,
      data: r.data
    }))
    .catch(e => ({
      success: false,
      error: `${onErrorPrefix}: ${e.response.data}`
    }));
}
