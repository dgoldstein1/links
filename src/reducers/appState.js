import { SET_LOADING, UPDATE_VIEW, SET_FATAL_ERROR } from "../actions/appState";

const initialState = {
  loading: true, // the app is / isn't loading
  fatalError: undefined, // error which  causes app to not load
  view: "splash" // one of "explore", "path", "splash", "about", "settings"
};

const visitReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_FATAL_ERROR:
      return Object.assign({}, state, {
        fatalError: action.fatalError
      });
    // set when the app is loading
    case SET_LOADING:
      return Object.assign({}, state, {
        loading: action.loading === undefined ? false : action.loading
      });
    // update where the user is currently looking
    case UPDATE_VIEW:
      return Object.assign({}, state, {
        view: action.view
      });
    default:
      return state;
  }
};

export default visitReducer;
