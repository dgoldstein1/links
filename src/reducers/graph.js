import _ from "lodash";

const initialState = {
  rootNode: {}, // root node of graph (what is in search or 'start from')
  selectedNode: {
    // node currently in view
    loading: true,
    error: "",
    description: "",
    img: "",
    node: {}
  },
  targetNode: {}, // target node in path ('ending at..')
  graph: {
    nodes: [],
    edges: []
  },
  maxShortestPaths: 5,
  directedShortestPath: true,
  loading: false,
  error: undefined,
  layout: "hierarchy",
  maxNeighbors: 15,
  topInfo: {
    betweenessEdges: [],
    betweenessNodes: [],
    pageRank: []
  }
};

const graphReducer = (state = initialState, action) => {
  switch (action.type) {
    case "CLEAR_GRAPH":
      return Object.assign({}, state, {
        graph: {
          nodes: [],
          edges: []
        }
      });
    case "SET_DIRECTED_SHORTEST_PATH":
      return Object.assign({}, state, {
        directedShortestPath: action.directedShortestPath
      });
    case "SET_MAX_NEIGHBORS":
      return Object.assign({}, state, {
        maxNeighbors: action.maxNeighbors
      });
    case "SET_GRAPH_ERROR":
      return Object.assign({}, state, {
        error: action.error
      });
    case "SET_GRAPH_LAYOUT":
      return Object.assign({}, state, {
        layout: action.layout
      });
    case "SET_GRAPH_LOADING":
      return Object.assign({}, state, {
        loading: action.loading
      });
    case "SET_ROOT_NODE":
      return Object.assign({}, state, {
        rootNode: _generateRoot(action.node.label, action.node.id)
      });
    case "SET_MAX_SHORTEST_PATHS":
      return Object.assign({}, state, {
        maxShortestPaths: action.maxShortestPaths
      });
    case "SET_ALL_PATHS_UNIQUE":
      return Object.assign({}, state, {
        pathsAreUnique: action.pathsAreUnique
      });

    case "SET_SELECTED_NODE":
      return Object.assign({}, state, {
        selectedNode: {
          ...state.selectedNode,
          node: action.node
        }
      });
    case "SET_SELECTED_NODE_INFO":
      return Object.assign({}, state, {
        selectedNode: {
          ...state.selectedNode,
          description: action.description,
          img: action.img,
          loading: action.loading,
          error: action.error,
          centrality: action.centrality
        }
      });
    case "SET_TARGET_NODE":
      return Object.assign({}, state, {
        targetNode: action.node
      });

    case "SET_GRAPH_PATH":
      // create array of all nodes from entries
      let nodes = action.entries.map(e => ({
        id: e.value,
        label: e.key,
        size: 1,
        color: "red"
      }));
      // create edges from each path in graph
      let edges = [];
      let entriesReverseMap = {};
      action.entries.forEach(e => (entriesReverseMap[e.value] = e.key));
      action.paths.forEach(p => {
        p.forEach((id, i) => {
          // dont add edge if we've reached the end
          if (i === p.length - 1) return;
          // lookup this entry and next entry
          edges.push({
            id: `${id}->-->--${p[i + 1]}`,
            source: id,
            target: p[i + 1],
            color: "red"
          });
        });
      });
      return Object.assign({}, state, {
        graph: { nodes, edges }
      });
    case "ADD_NEIGHBORS_TO_GRAPH":
      // set initial position to source
      action.neighbors = action.neighbors.map(n => ({
        ...n,
        size: 1
      }));
      // get list of edges to add
      let edgesToAdd = [];
      action.neighbors.forEach(n => {
        if (!n.id) console.log(n);
        edgesToAdd.push({
          id: `${action.node.id}->-->--${n.id}`,
          source: action.node.id,
          target: n.id,
          color: "black"
        });
      });
      // add in node to make sure it's in list of nodes
      action.neighbors.push(action.node);
      // return state
      return Object.assign({}, state, {
        graph: {
          nodes: _.unionBy(state.graph.nodes, action.neighbors, "id"),
          edges: _.unionBy(state.graph.edges, edgesToAdd, "id")
        }
      });
    default:
      return state;
  }
};

export function _generateRoot(label, id) {
  return {
    label,
    id,
    x: 0,
    y: 0,
    size: 1
  };
}

export default graphReducer;
