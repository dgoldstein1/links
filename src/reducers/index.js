// redux utils
import { combineReducers, createStore, applyMiddleware } from "redux";
import { createLogger } from "redux-logger";

// reducers in this directory
import appState from "./appState";
import graph from "./graph";

const Reducer = combineReducers({
  appState,
  graph,
});

const logger = createLogger({
  collapsed: true,
  diff: true,
});

let middleware = {};
if (process.env.REACT_APP_ENV === "development") {
  middleware = applyMiddleware(logger);
}

const store = createStore(Reducer, middleware);

export { store };
