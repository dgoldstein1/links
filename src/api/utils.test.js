import { makeRequest } from "./utils";
import moxios from "moxios";
import { store } from "../reducers";
import sinon from "sinon";
import axios from "axios";

describe("api utils", () => {
  describe("make request", () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });
    let testTable = [
      {
        name: "normal GET request",
        method: "get",
        url: `/test`,
        onErrorPrefix: "error getting ...",
        responseCode: 200,
        data: { test: "test" }
      },
      {
        name: "normal POST request",
        method: "post",
        bodt: { test: "test" },
        url: `/test`,
        onErrorPrefix: "error getting ...",
        responseCode: 200,
        data: { test: "test" }
      },
      {
        name: "GET request with response code 500",
        method: "get",
        url: `/test`,
        onErrorPrefix: "error getting ...",
        responseCode: 500,
        data: { test: "test" }
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        let onFulfilled = sinon.spy();
        moxios.stubRequest(t.url, {
          status: t.responseCode,
          response: t.data
        });
        makeRequest({ ...t }).then(r => {
          expect(r).toMatchSnapshot();
          done();
        });
      });
    });
  });
});
