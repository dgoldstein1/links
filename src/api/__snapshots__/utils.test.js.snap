// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api utils make request GET request with response code 500 1`] = `
Object {
  "error": "error getting ...: Request failed with status code 500",
  "success": false,
}
`;

exports[`api utils make request normal GET request 1`] = `
Object {
  "data": Object {
    "test": "test",
  },
  "success": true,
}
`;

exports[`api utils make request normal POST request 1`] = `
Object {
  "data": Object {
    "test": "test",
  },
  "success": true,
}
`;
